cl/recap/mergers.py

```diff
@@ -573,29 +573,51 @@ def merge_unnumbered_docket_entries(
 
     # Look for docket entries that match by equal long description or if the
     # long description is not set.
-    matched_entries = [
-        de.pk
-        for de in des
-        if de.description == docket_entry["description"] or not de.description
-    ]
-    if matched_entries:
+    matched_entries_queryset = des.filter(
+        Q(description=docket_entry["description"]) | Q(description="")
+    )
+    if await matched_entries_queryset.aexists():
         # Return the entry that matches the long description and remove the
         # rest if there are any duplicates.
-        matched_entries_queryset = des.filter(pk__in=matched_entries)
-        winner = matched_entries_queryset.earliest("date_created")
-        matched_entries_queryset.exclude(pk=winner.pk).delete()
+        winner = await matched_entries_queryset.aearliest("date_created")
+        await matched_entries_queryset.exclude(pk=winner.pk).adelete()
         return winner
 
     # No duplicates found by long description, choose the earliest as the
     # winner; delete the rest
-    winner = des.earliest("date_created")
-    des.exclude(pk=winner.pk).delete()
+    winner = await des.aearliest("date_created")
+    await des.exclude(pk=winner.pk).adelete()
     return winner
 
 
-def get_or_make_docket_entry(
+@sync_to_async
+def add_create_docket_entry_transaction(d, docket_entry):
+    with transaction.atomic():
+        Docket.objects.select_for_update().get(pk=d.pk)
+        try:
+            de, de_created = DocketEntry.objects.get_or_create(
+                docket=d, entry_number=docket_entry["document_number"]
+            )
+        except DocketEntry.MultipleObjectsReturned:
+            logger.error(
+                "Multiple docket entries found for document "
+                "entry number '%s' while processing '%s'",
+                docket_entry["document_number"],
+                d,
+            )
+            duplicate_de_queryset = DocketEntry.objects.filter(
+                docket=d, entry_number=docket_entry["document_number"]
+            )
+            de = duplicate_de_queryset.latest("date_created")
+            duplicate_de_queryset.exclude(pk=de.pk).delete()
```

Comment:
I'm not sure. I'm about 95% certain it's OK, but do you know if we still have the queries we used during that last analysis?

Reply:
Well, it's not OK. I just took a sample using the query listed here: 

https://github.com/freelawproject/courtlistener/issues/2144#issuecomment-1198516980

But tweaked to do a 1% sample:

```
courtlistener=> SELECT docket_id, entry_number, count(*) 
FROM public.search_docketentry 
TABLESAMPLE SYSTEM(1) WHERE entry_number is not null 
GROUP BY docket_id, entry_number 
HAVING count(*) > 1 LIMIT 100;
```

That gave me 100 results, which I made into a little script that just said:

```
firefox 'https://www.courtlistener.com/docket/5701473/d/?entry_gte=302&entry_lte=302&order_by=desc'
firefox 'https://www.courtlistener.com/docket/48828617/d/?entry_gte=51&entry_lte=51&order_by=desc'
firefox 'https://www.courtlistener.com/docket/8484552/d/?entry_gte=35&entry_lte=35&order_by=desc'
firefox 'https://www.courtlistener.com/docket/15660936/d/?entry_gte=4&entry_lte=4&order_by=desc'
[...× 100, all the way down...]
```

I opened all of those, and out of them, a few looked suspicious. Of those, a few actually have duplicates that are correct. PACER shows the duplicates too. For example, this one:

https://www.courtlistener.com/docket/53634944/d/?entry_gte=15&entry_lte=15&order_by=desc

In PACER, shows:

![image](https://github.com/freelawproject/courtlistener/assets/236970/b27f50f6-f8b2-4b0e-82f6-4997f2e5d9a8)

And here are two more:
 - https://www.courtlistener.com/docket/7573403/d/?entry_gte=97&entry_lte=97&order_by=desc
 - https://www.courtlistener.com/docket/6754636/d/?entry_gte=19&entry_lte=19&order_by=desc

I also noticed that in a few cases, you would have a `docket_entry.description` value in one docket entry, while each duplicate had the recap_document.description field. In that case, if you delete the wrong one, you'll delete the docket_entry.description which is only stored on one of the items. That's not great.

So....I think what I'd do, if we want to pursue this, is:
 - Check if one of the two+ items has the recap_document.description value. If so, it's the one to keep. Then...
 - Delete the duplicate the recap_document.description is the same on all.

I think that'd solve the two problems:
 - We want to keep the docket entry description if we have it on only one of the dups.
 - We don't want to delete if content is different, because those could be valid duplicates.

Reply:
Hmm, is the sequence number the same for each duplicate entry? Thinking we could maybe add a constraint which applies to `("docket", "entry_number", "recap_sequence_number")` or something like that? Do any of these duplicates have the same sequence number?

Reply:
We don't always have the sequence number, but you're right that if they have different `pacer_doc_id` values.

Reply:
> We don't always have the sequence number

Well a unique constraint wouldn't get enforced for an entry with a null `pacer_sequence_number` AFAIU. Maybe tweak the query [here](https://github.com/freelawproject/courtlistener/issues/2144#issuecomment-1198516980) and see if there are currently any duplicate sequence numbers?

Reply:
Why use the sequence number instead of the pacer_doc_id?

Reply:
It doesn't look like `pacer_doc_id` is part of the `DocketEntry` model while sequence number is.

Reply:
Gotcha. I wouldn't trust sequence number though. I think we often lack it in any case, limiting its utility. I did just run the query you suggested, and it showed some dups, but I didn't have time to really dig into them (see below). 

How about using description, though? 

```
courtlistener=> SELECT docket_id, pacer_sequence_number, count(*) 
FROM public.search_docketentry 
TABLESAMPLE SYSTEM(1) WHERE entry_number is not null and pacer_sequence_number is not null 
GROUP BY docket_id, pacer_sequence_number 
HAVING count(*) > 1 LIMIT 100;
 docket_id | pacer_sequence_number | count 
-----------+-----------------------+-------
   7896133 |                   767 |     2
   5046011 |                   144 |     2
   4538381 |                327941 |     2
   5286768 |               2119735 |     2
   8472234 |                    90 |     2
   5008358 |                   655 |     2
   5008358 |                   644 |     2
  16229680 |                   146 |     2
   5943323 |                    84 |     2
   5008358 |                   642 |     2
   7908978 |                    53 |     2
   5954916 |                   254 |     2
   5701473 |                  1351 |     2
  17371345 |                   433 |     2
  17371345 |                   412 |     2
  60025217 |                    15 |     2
   5286768 |               2119772 |     2
  13186126 |                   152 |     2
   4538381 |                327939 |     2
   5008358 |                   653 |     2
   5008358 |                   639 |     2
   5008358 |                   649 |     2
   4561818 |                   435 |     2
   5008358 |                   646 |     2
   4538381 |                327935 |     2
   5286768 |               2119738 |     2
   6449749 |                    22 |     2
   4538381 |                327937 |     2
   5701473 |                  1685 |     2
   5589640 |                    69 |     2
   6705948 |                   404 |     2
  16229680 |                   150 |     2
   4215336 |                  2854 |     2
  17371345 |                   430 |     2
   5286768 |               2119740 |     2
   5286768 |               2119750 |     2
   4216072 |                   287 |     2
  17371345 |                   435 |     2
   4130359 |                    19 |     2
   5286768 |               2119774 |     2
   5286768 |               2119754 |     2
  17371345 |                   416 |     2
  13381934 |               1765606 |     2
  67309034 |                    13 |     2
  13518597 |                   222 |     3
   7035785 |                    67 |     2
   4581945 |                   172 |     2
   5986115 |                  3367 |     2
   4395118 |                    62 |     2
   5842789 |                   100 |     3
   4708646 |               3476586 |     5
```

Reply:
> I think we often lack it in any case, limiting its utility.

Well we'll still need a fallback, but if we have an actual unique identifier here it helps with the deduplication logic I think.

> I did just run the query you suggested, and it showed some dups, but I didn't have time to really dig into them (see below).

Well good news seems to be that these appear to be duplicates only in our database, so it looks like uniqueness may actually be a guarantee we can rely on at least within each docket.

> How about using description, though?

I don't think that's guaranteed to be unique.

Reply:
I think I've got this refactored now to use the pacer sequence numbers to deduplicate docket entries and properly handle duplicate `document_number` numbers as long as a sequence number is present in the docket being uploaded.

Reply:
That sounds good. @albertisfu want to take another look, please?

Reply:
This seems good, I only found bug in the case `pacer_seq_no is None`, one comment below.